CIP-AnyDataIsDatum/README.md

```diff
@@ -0,0 +1,45 @@
+---
+CIP: ????
+Title: Hash-Checked Data
+Author: Maksymilian 'zygomeb' Brodowicz <zygomeb@gmail.com>
+Status: Draft
+Type: Standards
+Created: 2022-10-25
+License: CC-BY-4.0
+---
+
+## Simple Summary / Abstract
+
+Hashing data on-chain is quite expensive from a computational perspective. Currently, however, we are forced to use the `serializeData` primitive combined with a hashing function of our choice. It would be computationally beneficial to let the node hash the data before running the smart contracts, giving us only a guarantee that this mapping is correct, just as we do with Datums presently.
+
+Therefore we propose extending the ScriptContext from just a mapping between DatumHash and Datum to a more universal mapping between BuiltinByteString and BuiltinData.
+
+## Motivation / History
+
+We can repeat the same motivation as [CIP 42](https://cips.cardano.org/cips/cip42/), and more broadly emphasizing the want to leverage complex data structures on chain. 
+
+When presented with either a datum inside a datum or in general, data behind a hash, we are forced to spend a lot of computational resources to first serialize and then hash and then check for equality. This comes up quite often when implementing counter-double-satisfation measures (payment to script double satisfied).
+
+## Specification
+
+### Extending the ScriptContext
+
+We change the field `txInfoData` to `[(BuiltinByteString, BuiltinData)]`. Nominally, this changes nothing as the `DatumHash` and `Datum` types are equivalent to these. Importantly however, we allow any pair to be pushed to this list, and the node will hash the `BuiltinData` to verify that it matches the `BuiltinByteString`.
+
+It is worth noting that we can keep the types as they are now.
+
+## Rationale
+
+Just as we use nodes to be the source of truth on hash equality for datums we should encourage the use of them for more data, if it can be known statically.
+
+## Backwards compatibility
+
+The change can be made to the node without a creating a new language version as it does not affect the past transactions.
```

Comment:
This is false. It would prevent old scripts from evaluating. It would at least require a hard fork and a change on a protocol version and probably a new language version.

Reply:
I thought this wasn't the case as transactions in the past simply wouldn't have any additional pairs, and ones in the future could have them. This means that a node version that supports this change would be able to work with both at the same time without the need to explicitly state when the ability to use both was introduced. 

The only concern we could have is that a script explicitly relies on only datums being present which to that thinking abou it now I say fair. This should have a new version for the sake of this hypothetical script. I shall correct this.

Reply:
@zygomeb it does indeed require a hard fork, because only after a hard fork can we assume that everyone is running software that agrees to what is allowed.

> a node version that supports this change would be able to...

right, but a node version that does note support this change will disagree with the upgraded nodes that take advantage of the change.

Reply:
I see -- that is indeed true. In that case this should be packaged with a hard fork, good thing to confirm.